### ISSUE TITLE
Rebuild main menu as full SwingUI

### ISSUE THREAD

**Issue by andreas-eberle** [jsettlers/settlers-remake#268]
Currently the main menu is rendered with custom UI elements build with OpenGL. However, for the main menu, we need a proper UI with scrollable lists, tabs... 

Therefore this ticket targets to create a new main menu build as complete Swing UI. (This is a follow up of #141; however, we will use Swing instead of JavaFX).

This also includes #140.

**Comment by andreasb242** [comment]
I'm working at the Look & Feel. See LookAndFeelSample.java as sample.

The Look and Feel has to be installed like this:
```java
			for (LookAndFeelInfo info : UIManager.getInstalledLookAndFeels()) {
				if ("Nimbus".equals(info.getName())) {
					UIManager.setLookAndFeel(info.getClassName());
					break;
				}
			}

			UIManager.addAuxiliaryLookAndFeel(new SettlerLookAndFeel());
```

The main concept: I use Nimbus as Default Look and Feel, so generally all Components look like Nimbus.

Then each component is configured for a specific Look and Feel, e.g. for Labels I create 3.

```java
		JButton button1 = new JButton("Button1");
		button1.putClientProperty(LFStyle.KEY, LFStyle.BUTTON_MENU);
		backgroundPanel.add(button1);

		// ********************************************************
		// IMPORTANT
		// Update tree ui after the style of all components is set
		// ********************************************************
		SwingUtilities.updateComponentTreeUI(backgroundPanel);
		backgroundPanel.invalidate();
		// ********************************************************
```
Currently nothing is rescaled except the Panel background.
So if a button text is to long, there is a layout problem. I will look at this if there is really a problem.

As log as there is no special style for a Component, simple use the default Look and feel, we can later add styles for all components we need.

And how the example looks now:
![s3-controls](https://cloud.githubusercontent.com/assets/822958/12153599/f7991ea4-b4ba-11e5-91aa-9265f800013a.png)

**Comment by codingberlin** [comment]
Current state:
- you can open a new single player game and set the AIs, your map slots, your team, your amount of start resources (does influence original maps only), reduce the number of players and start the game
- you can create a multiplayer game, which will be displayed in the old swing UI

ToDo:
- manage joining the multiplayer game
- display joined players in the host multiplayer panel
- start the multiplayer game
- chat
- merge the look and feel from @andreasb242

**Comment by codingberlin** [comment]
@andreas-eberle, @michaelzangl When I quit the game within the new GUI with q I often get the following error message.
I guess I need to shutdown something to avoid that error. Do you have an idea?
```
Executing SyncTaskPacket(lockstep: 2602 tasks: [SimpleGuiTask: BUILD playerId: 2]) in lockstep: 2602 (game time: 260200ms / 00:04:20:200)
executeTask(GuiTask): BUILD
Executing SyncTaskPacket(lockstep: 2602 tasks: [SimpleGuiTask: BUILD playerId: 3]) in lockstep: 2602 (game time: 260200ms / 00:04:20:200)
executeTask(GuiTask): BUILD
Q
action(Action): EXIT   at game time: 262950
Endgame statistic:
Player 0: amountOfProducedSoldiers: 0, amountOfProducedGold: 0, amountOfProducedMana: 0
Player 1: amountOfProducedSoldiers: 0, amountOfProducedGold: 0, amountOfProducedMana: 0
Player 2: amountOfProducedSoldiers: 0, amountOfProducedGold: 0, amountOfProducedMana: 0
Player 3: amountOfProducedSoldiers: 0, amountOfProducedGold: 0, amountOfProducedMana: 0
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007fff86c470dd, pid=703, tid=1811
#
# JRE version: Java(TM) SE Runtime Environment (8.0_65-b17) (build 1.8.0_65-b17)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.65-b01 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libobjc.A.dylib+0x10dd]  objc_msgSend+0x1d
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/stephan/Documents/settlers-remake/jsettlers.main.swing/hs_err_pid703.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
#

Process finished with exit code 134
```

**Comment by codingberlin** [comment]
@andreas-eberle What do you think?
Should I enhance the multiplayer protocol  to send changes in the slots, teams, AIs in an open game or should I finish this issue without that in order to review and integrate the UI fast and add this in a second iteration?
I could argue for both ways.

**Comment by michaelzangl** [comment]
I think that the cleanest solution to do this would be to add the new network code in a new branch.

It is more difficult to test but if you feel up to writing test cases for that, you can simply create a new branch with the networking code enhancements and test cases. We can then merge that branch in master independent from the new UI and I can start working on the Android side.

**Comment by andreas-eberle** [comment]
Regarding the crash: I will need to look into this more closely. I guess, it is "caused" by a change of mine. Since that change, the logic cleans up its stuff correctly, but the graphics sometimes tends to try to render after the cleanup and crashes then. However, I've never seen a SIGSEGV caused by that, only "ignorable" exceptions...

Regarding the network changes: I also think this should be extended in a different branch. I think it's best to have the UI first (but probably disable reordering etc. for multiplayer) and implement that feature seperately. Smaller changes makes it easier to review them and therefore faster to merge. ;)

**Comment by codingberlin** [comment]
Ok I will leave Network changes out of this Issues and do them in another Issue and branch

**Comment by andreasb242** [comment]
@codingberlin Can you look at the hs_err_pid*.log?
Usually you see the cause of the VM crash in this Logfile...

**Comment by andreas-eberle** [comment]
If you rename the file from *.log to *.txt, you should be able to attach it here. (Github only allows .txt files, not .log files...)

**Comment by codingberlin** [comment]
Today it happened again. Here is the hs_err_pid as txt
[hs_err_pid887.txt](https://github.com/jsettlers/settlers-remake/files/101479/hs_err_pid887.txt)

**Comment by michaelzangl** [comment]
AWT-AppKit? This is bad.

Last unusual thing seems to have been a NPE in jsettlers.logic.objects.ProgressingObject.getStateProgress().

**Comment by andreasb242** [comment]
May this Pull Request is already the fix for this Bug?
https://github.com/sgothel/jogl/pull/79/files?diff=split

The important part of the Stacktrace:
```
C  [libsystem_platform.dylib+0x1405]  _platform_memmove$VARIANT$Nehalem+0xc5
C  [libobjc.A.dylib+0x9317]  _memdup_internal+0x31
C  [libobjc.A.dylib+0x9197]  _ZL15fixupMethodListP13method_list_tbb+0x2d
C  [libobjc.A.dylib+0x19d51]  _ZL17attachMethodListsP10objc_classPP13method_list_tibbb+0x152
C  [libobjc.A.dylib+0x14547]  _ZL12realizeClassP10objc_class+0xaff
C  [libobjc.A.dylib+0x176fd]  lookUpImpOrForward+0xc8
C  [libobjc.A.dylib+0x11ac]  objc_msgSend+0xec
C  [libnativewindow_macosx.jnilib+0x5127]  Java_jogamp_nativewindow_jawt_macosx_MacOSXJAWTWindow_UnsetJAWTRootSurfaceLayer0+0xe7
j  jogamp.nativewindow.jawt.macosx.MacOSXJAWTWindow.UnsetJAWTRootSurfaceLayer0(JJ)V+0
```

_platform_memmove is the point where it crashed, but for sure not the reason...

**Comment by michaelzangl** [comment]
memmove is simply moving memory out of range. One problem seems to be that graphics is still rendering while logic is shutting down.

We can work around this by avoiding it. The JOGL PR might fix that. But the JOGL guys are not really good at maintaining it, so it will take a lot of time until it is available.

**Comment by codingberlin** [comment]
![bildschirmfoto 2016-01-22 um 23 47 58](https://cloud.githubusercontent.com/assets/6422616/12525261/6525b4e6-c163-11e5-8b65-bf8a93561c19.png)

The current state:
- you can open, join and start multiplayer games
- in single player you can change everything (slots, AI, number of players, ...)
- in multiplayer games you can't change anything because this is missing in the network protocol and will be implemented in another issue
- you can send chat messages

ToDo:
- add ´´´SwingUtilities.invokeLater(...´´´everywhere it is needed
- add the look and feel from @andreasb242 (buttons, flexible labels, chat styling, Ready-Button, Toggle-Buttons for the MainMenu)

**Comment by andreas-eberle** [comment]
@codingberlin: Wow, this looks great already!

**Comment by andreas-eberle** [comment]
Fixed with e6b58eb.
